tezedge_state/src/chunking/message_read_buffer.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use std::io::{self, Read};
+
+use tezos_messages::p2p::{
+    binary_message::{BinaryRead, SizeFromChunk},
+    encoding::peer::PeerMessageResponse,
+};
+
+use super::{ChunkReadBuffer, ReadMessageError};
+use crate::PeerCrypto;
+
+/// Read buffer for connected peer(`PeerMessage`).
+///
+/// This message can be split into multiple chunks. Encrypted part of
+/// the first chunk includes information (`4 bytes`) about whole message's
+/// size. We need to process chunks until decrypted accumulated message
+/// isn't of that size.
+#[derive(Debug, Clone)]
+pub struct MessageReadBuffer {
+    message_buf: Vec<u8>,
+    message_len: usize,
+    /// Bytes read for current message.
+    read_bytes: usize,
+    chunk_reader: ChunkReadBuffer,
+}
+
+impl MessageReadBuffer {
+    pub fn new() -> Self {
+        Self {
+            message_buf: vec![],
+            message_len: 0,
+            read_bytes: 0,
+            chunk_reader: ChunkReadBuffer::new(),
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.message_len > 0 && self.message_buf.len() >= self.message_len
+    }
+
+    /// Returns if more might be available to be read.
+    fn _read_from<R: Read>(
+        &mut self,
+        reader: &mut R,
+        crypto: &mut PeerCrypto,
+    ) -> Result<PeerMessageResponse, ReadMessageError> {
+        loop {
+            self.chunk_reader.read_from(reader)?;
+            // TODO: stop reading if message_len < message_buf.len() + chunk_expected_len
```

Comment:
@binier 
still relevant comment?

Reply:
That's an optimization maybe good to have (unless peer's implementation is broken, that optimization should never apply), for now I don't want to break anything so it can be left as is.

Reply:
@binier 
ok, cool, so please, just update that TODO with this comment, you just wrote here,
that we now what to do with that later